More input handling options

fluffysquirrels opened the pull request:
In particular:

* Arguments to read to the end of a file or follow the end (like tail -f)
* Extract a chosen stimulus port instead of only 0

Also:
* Logging with env_logger

japaric commented:
Thanks, @fluffysquirrels. This LGTM.

@homunkulus r+

homunkulus commented:
:pushpin: Commit ba95d35 has been approved by `japaric`

<!-- @homunkulus r=japaric ba95d35bc343f714984f94297e2828490ae00efa -->

homunkulus commented:
:hourglass: Testing commit ba95d35bc343f714984f94297e2828490ae00efa with merge f62f22c8a2422b2e56168813145f8e079582677d...

homunkulus commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/japaric/itm/build/1.0.7)

fluffysquirrels commented:
The appveyor build failed with this message:

```
error[E0554]: #![feature] may not be used on the stable release channel
 --> src\bin\itmdump.rs:2:1
  |
2 | #![feature(conservative_impl_trait)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
error: aborting due to previous error
```

It looks like travis and appveyor are both building with Rust stable. If you have no strong preference for using stable I'd suggest switching to a different toolchain. Perhaps just `nightly` or `beta`? I could also re-write the patch to work on stable if required.

fluffysquirrels commented:
I've just pushed some more changes. One of them is to try the CI builds with `nightly` instead of `stable`, which hopefully will fix the build error. Can you try the builds again?

Also:
* Split packet decoding into a library (completes issue <https://github.com/japaric/itm/issues/1>)
* Added packet decoding unit tests.
* Fixed /tests/smoke.

fluffysquirrels commented:
I tried triggering a Travis build myself on my branch, but that failed. I've opened a [Github issue with travis-ci](https://github.com/travis-ci/travis-ci/issues/8956).

An [AppVeyor build](https://ci.appveyor.com/project/fluffysquirrels/itm/build/1.0.1) is running for my branch now.

japaric commented on src/bin/itmdump.rs:
```diff
@@ -57,93 +60,95 @@ fn main() {
+    // Unreachable.
+}
+
+fn open_read<'a>(matches: &ArgMatches) -> Result<impl io::Read + 'a> {
```
Since you are creating a *boxed* trait object in both branches I think the return type should be `Box<Read + 'static>` instead of `impl Read`. Then the `impl_conservative_trait` feature gate can be dropped.

fluffysquirrels replied:
Fixed. Yes, that was the fix I'd imagined.

japaric commented on src/packet.rs:
```diff
@@ -0,0 +1,49 @@
+}
+
+/// Contents of an Instrumentation packet, with data from a software application
+pub struct Instrumentation {
```
Personally, I prefer to keep fields private and provide accessor methods to the fields (e.g. `fn payload(&self) -> &[u8]`). I go for public fields when I want to let the user create an instance of the struct. But I don't think we want to allow that in this case; also I think that as it is this struct lets the user construct invalid `Instrumentation` instances (port can only be in the range 0-31 iirc).

fluffysquirrels replied:
Fixed.

japaric commented on src/packet.rs:
```diff
@@ -0,0 +1,49 @@
+
+/// Represents a complete received packet.
+#[derive(Debug)]
+pub struct Packet {
```
Same comment as the `Instrumentation` one. The user can construct `Packet`s with invalid header values, I think.

fluffysquirrels replied:
Fixed. I think I also prefer accessor methods here, but because that allows us to change the internal representation freely.

japaric commented on src/decoder.rs:
```diff
@@ -0,0 +1,181 @@
+}
+
+/// Decode a single packet from a slice of bytes.
+pub fn from_slice(s: &[u8]) -> Result<Packet> {
```
Not really important but is it possible to return a `Result` type here that doesn't include the `Io` variant? I think that no I/O error can occur in this function so it would be nice to reflect that in the signature. I'm not sure if error-chain allows such thing though; I have never used the error-chain! macro with the `errors` fields.

fluffysquirrels replied:
Won't change, as discussed.

fluffysquirrels commented:
Thanks for the review. I agree with the comments and will make the changes
over the next day or so.

error_chain! generates an enum for ErrorKind with all your forwarded and
local errors as variants, so no, it's not simple to show in the type
signature that no io::Error can occur here. I would leave it as is.

japaric commented:
> will make the changes over the next day or so

Thanks!

>  I would leave it as is

yeah, that's fine.

fluffysquirrels commented:
My latest commit applies the changes you suggested in the review.

japaric commented:
Thanks again, @fluffysquirrels. This looks great.

@homunkulus r+

homunkulus commented:
:pushpin: Commit fe296a8 has been approved by `japaric`

<!-- @homunkulus r=japaric fe296a8863ce9de33d4bbcd60249f88e2ee14b81 -->

homunkulus commented:
:hourglass: Testing commit fe296a8863ce9de33d4bbcd60249f88e2ee14b81 with merge a8759c77d4eee208211da16dddc6847ac094c814...

homunkulus commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/itm/builds/320706047?utm_source=github_status&utm_medium=notification)

japaric commented:
Those build errors are due to missing C cross toolchain bits in the Travis environment. The easiest way to fix the errors is to use Cross. I have opened #12 to switch the CI stuff to Trust, and thus to use Cross. Unfortunately that PR will break this one and you'll have to rebase ... Sorry for the trouble.

homunkulus commented:
:umbrella: The latest upstream changes (presumably #12) made this pull request unmergeable. Please resolve the merge conflicts.

fluffysquirrels commented:
@japaric Rebase done. I didn't do anything meaningful with the CI config so the rebase was trivial. This built fine locally.

fluffysquirrels commented:
The Travis build on my branch failed everywhere because it tried to use rust stable, and itm requires heapless, which requires untagged-option, which requires #![feature(untagged_unions)], which requires nightly. I'll make a quick change to stop using heapless.

fluffysquirrels commented:
Travis build looks good on my branch: <https://travis-ci.org/fluffysquirrels/itm/builds/320830868>

japaric commented:
Thanks, @fluffysquirrels. It's great to support stable.

@homunkulus r+

homunkulus commented:
:pushpin: Commit feba5a0 has been approved by `japaric`

<!-- @homunkulus r=japaric feba5a006914a3ae2ad52165709bdacd47c48ae0 -->

homunkulus commented:
:hourglass: Testing commit feba5a006914a3ae2ad52165709bdacd47c48ae0 with merge 19a77e01188765c77c9a4e6932bf7c9b4db12291...

homunkulus commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/japaric/itm/build/1.0.14), [status-travis](https://travis-ci.org/japaric/itm/builds/321062640?utm_source=github_status&utm_medium=notification)
Approved by: japaric
Pushing 19a77e01188765c77c9a4e6932bf7c9b4db12291 to master...
